DSP Accel Library/Testing and Images/TestingLibraryScripts.m

```diff
@@ -0,0 +1,129 @@
+%% TESTING ONLY, USER DEFINED VARS
+clear;
+TableFn = "log(X_in)";
+maxInput = 1;
+floor_in_tab = 2^-12;
+error_cap_tab = 0.001;
+ERR_DIAG = true;
+N_bits_tab = 4;
+M_bits_tab = 5;
+igotthis = false;
+W_bits = 32;
+F_bits = 28;
+
+% things to display: table size, floor input?, max input, accuracy
+% things to figure out: M_bits, N_bits, Min_val, X_in table, Y_out table,
+% max_error, ram_size, table_size 
+
+d = ceil(log2(maxInput));
+if(igotthis)
+    N_bits = N_bits_tab;
+    M_bits = M_bits_tab;
+else
+    N_bits = ceil(log2(d-log2(floor_in_tab)+1));
+    % M_bits needs to be defined later, as it must be just enough to meet
+    % accuracy standards defined by user, with linear interpolation
+    M_bits = 1; %this will be updated later as needed
+end;
+
+repeatFlag = true;
+while(repeatFlag)
+    % Define a N point log2 spaced range of inputs, from 2^(d-(2^N_bits -1)) to almost 2^(d+1), 
+    % this effectively covers the user defined input range, down to a floor
+    % value.
+    X_in = zeros(1, 2^(M_bits+N_bits));
+    addr = 1;
+    for NShifts = 2^N_bits-1:-1:0
+        for M = 0:2^M_bits - 1
+            X_in(addr) = 2^(d-NShifts) + M*2^(d-NShifts-M_bits);
+            addr = addr+1;
+        end
+    end
+
+    % use function to define output. note: function must contain input X_in
+    % within the string to work properly, and the only output of the function
+    % must be Y_out.
+    Y_out = eval(TableFn);
+
+    ram_size = N_bits+M_bits;
+    Table_init = Y_out;
+    max_val = 2^d+ (2^M_bits -1)*2^(d-M_bits);
+    min_val = 2^(d-(2^N_bits-1));
+    %set_param(gcb,'MaskDisplay',"disp(sprintf('Programmable Look-up Table\nMemory Used = %d samples and coeffs\nClock Rate Needed = %d Hz', FIR_Uprate*2, Max_Rate)); port_label('input',1,'data'); port_label('input',2,'valid'); port_label('input',3,'Wr_Data'); port_label('input',4,'Wr_Addr'); port_label('input',5,'Wr_En'); port_label('output',1,'data'); port_label('output',2,'valid'); port_label('output',3,'RW_Dout');")
+
+    %%%%%%%%%% check and identify error %%%%%%%%%%%
+    % Make some values for an "ideal" lookup table with log spaced points
+    X_in_Ideal = logspace(log10(min_val), log10(max_val), 2^(ram_size+2));
+    X_temp = X_in;
+    X_in = X_in_Ideal;
+    % identify the values of the function at X_in_Ideal
+    Y_out_Ideal = eval(TableFn);
+
+    % Get values for lookup table (already have X_in)
+    X_in = X_temp;
```

Comment:
This feels a little clunky to me with the temp variable.  Maybe just assign the X_in values to another variable and use that throughout?  Also, why is the eval function used here?  I think calling the log function directly would be a little cleaner

Reply:
It's because the user function can only be typed into a mask as a string, so I have to tell them to set an input to the function "xIn". I also have to use this same function with 2 different input sets: the table input set and the test input set. I could alter the string, but this seemed easier.